homeassistant/components/usb/__init__.py

```diff
@@ -0,0 +1,138 @@
+"""The USB Discovery integration."""
+from __future__ import annotations
+
+import dataclasses
+import datetime
+import logging
+import sys
+
+from serial.tools.list_ports import comports
+from serial.tools.list_ports_common import ListPortInfo
+
+from homeassistant import config_entries
+from homeassistant.const import EVENT_HOMEASSISTANT_STARTED, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.loader import async_get_usb
+
+from .flow import FlowDispatcher, USBFlow
+from .models import USBDevice
+from .utils import usb_device_from_port
+
+_LOGGER = logging.getLogger(__name__)
+
+# Perodic scanning only happens on non-linux systems
+SCAN_INTERVAL = datetime.timedelta(minutes=60)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the USB Discovery integration."""
+    usb = await async_get_usb(hass)
+    usb_discovery = USBDiscovery(hass, FlowDispatcher(hass), usb)
+    await usb_discovery.async_setup()
+    return True
+
+
+class USBDiscovery:
+    """Manage USB Discovery."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        flow_dispatcher: FlowDispatcher,
+        usb: list[dict[str, str]],
+    ) -> None:
+        """Init USB Discovery."""
+        self.hass = hass
+        self.flow_dispatcher = flow_dispatcher
+        self.usb = usb
+        self.seen: set[tuple[str, ...]] = set()
+
+    async def async_setup(self) -> None:
+        """Set up USB Discovery."""
+        if not await self._async_start_monitor():
+            await self._async_start_scanner()
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, self.async_start)
+
+    async def async_start(self, event: Event) -> None:
+        """Start USB Discovery and run a manual scan."""
+        self.flow_dispatcher.async_start()
+        await self.hass.async_add_executor_job(self.scan_serial)
+
+    async def _async_start_scanner(self) -> None:
+        """Perodic scan with pyserial when the observer is not available."""
+        stop_track = async_track_time_interval(
+            self.hass, lambda now: self.scan_serial(), SCAN_INTERVAL
+        )
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, callback(lambda event: stop_track())
+        )
+
+    async def _async_start_monitor(self) -> bool:
+        """Start monitoring hardware with pyudev."""
+        if not sys.platform.startswith("linux"):
```

Comment:
I think we can also skip that normal container installation because they don't get the new device mapping and need also the udev socket mapped into /run for get events with network=host.

Reply:
Do have a suggested way to detect that?

Reply:
`async_get_system_info` from `homeassistant.helpers.system_info`

Reply:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/homeassistant/system_health.py#L23-L24
Hassio will work but docker not (hassio work because we map the host dev into container  to support plug&play and mapping udev into /run)

Reply:
Thanks.  will adjust after https://github.com/home-assistant/core/pull/54953

Reply:
https://github.com/home-assistant/core/pull/54987